ui/proposals_page.go

```diff
@@ -168,81 +142,33 @@ func (pg *proposalsPage) Handle(common pageCommon) {
 		pg.wallet.SyncProposals()
 		common.refreshPage()
 	}
-}
-
-func (pg *proposalsPage) addDiscoveredProposal(proposal dcrlibwallet.Proposal) {
-	for i := range pg.tabs.tabs {
-		if pg.tabs.tabs[i].category == proposal.Category {
-			item := proposalItem{
-				btn:               new(widget.Clickable),
-				proposal:          proposal,
-				voteBar:           pg.theme.VoteBar(pg.infoIcon, pg.legendIcon),
-				infoIcon:          pg.infoIcon,
-				stateInfoTooltip:  pg.theme.Tooltip(),
-				stateTooltipLabel: pg.theme.Caption(""),
-			}
-			pg.tabs.tabs[i].proposals = append([]proposalItem{item}, pg.tabs.tabs[i].proposals...)
-			break
-		}
-	}
-}
-
-func (pg *proposalsPage) updateProposal(proposal dcrlibwallet.Proposal) {
-out:
-	for i := range pg.tabs.tabs {
-		for k := range pg.tabs.tabs[i].proposals {
-			if pg.tabs.tabs[i].proposals[k].proposal.Token == proposal.Token {
-				pg.tabs.tabs[i].proposals = append(pg.tabs.tabs[i].proposals[:k], pg.tabs.tabs[i].proposals[k+1:]...)
-				break out
-			}
-		}
-	}
-	pg.addDiscoveredProposal(proposal)
-}
 
-func (pg *proposalsPage) onFetchSuccess(proposals []dcrlibwallet.Proposal) {
-	for i := range proposals {
-		item := proposalItem{
-			btn:               new(widget.Clickable),
-			proposal:          proposals[i],
-			voteBar:           pg.theme.VoteBar(pg.infoIcon, pg.legendIcon),
-			infoIcon:          pg.infoIcon,
-			stateInfoTooltip:  pg.theme.Tooltip(),
-			stateTooltipLabel: pg.theme.Caption(""),
-		}
-
-		for k := range pg.tabs.tabs {
-			if pg.tabs.tabs[k].category == proposals[i].Category {
-				pg.tabs.tabs[k].proposals = append(pg.tabs.tabs[k].proposals, item)
-				break
+	select {
+	case prop := <-*pg.proposalSync:
+		if prop.ProposalStatus == wallet.Synced {
+			if !pg.proposalsItemSet {
+				pg.initializeProposaltabItems()
 			}
+			go pg.updateProposalState()
+			pg.isSynced = true
+		} else if prop.ProposalStatus == wallet.NewProposalFound {
+			pg.addDiscoveredProposal(*prop.Proposal)
+			common.refreshPage()
+		} else if prop.ProposalStatus == wallet.VoteStarted || prop.ProposalStatus == wallet.VoteFinished {
+			pg.updateProposal(*prop.Proposal)
+			common.refreshPage()
 		}
-	}
-	pg.onFetchComplete()
-	pg.state = categoryStateFetched
-}
-
-func (pg *proposalsPage) onFetchError(err error) {
-	pg.state = categoryStateError
-	pg.onFetchComplete()
-	pg.notify(err.Error(), false)
-}
-
-func (pg *proposalsPage) onFetchComplete() {
-	if !pg.hasFetchedInitialProposals {
-		pg.hasFetchedInitialProposals = true
+	default:
```

Comment:
There's no need for a default case if the case is empty

Reply:
Removing the default cases causes the app to freeze when proposals page is selected. I spent the last 30 mins trying to figure out this behavior.